README.md
# fqclean
Program: fqclean   filtrate/trim pair-end reads with adapter or with too many low-quality-base
Version: v2.0
Contact: yerui <yerui@genomics.cn>

Usage: fqclean in1.fq[.gz] in2.fq[.gz] out1.fq.gz out2.fq.gz

       -n [f]    read in which rate of N > [f] will be discarded. [0.1]
       -q [f]    read in which rate of low-quality-base > [f] will be discarded. [0.5]
       -c [c]    phred quality < [c] is treated as low quality base. [ 'F' ]
       -p [i]    [i] reads as a part to output. [0]
       -m [i]    minimum length of reads after cleaning. [30]
       -u [s]    UID types. ( NNNNNACT  means 5bp barcode + 3bp fix
                              ACGNNNNN  meads 3bp fix + 5bp barcode
                              NNNNNNNN  meads 8bp barcode. )
       -t [i]    trim first N bp of reads (if has UID, trim N bp after UID) [0]



v2.0/NEWS
Thu May  8 15:33:17 CST 2014    v2.0
----------------------------------------------------------------
1. add the option -t, to trim the head of reads.
2. discard reads which has mismatch on fix-seq of barcode.

changed files: UID_type.h fq_read.h main.cc main.h

Thu Apr 25 16:38:27 HKT 2013    v1.1
----------------------------------------------------------------
1. fix a major bug, which allow barcode contamination at the end of reads.

Wed Apr 10 21:39:30 CST 2013    v1.0
----------------------------------------------------------------
1. add -u option, fetch UID and put it into readID.

Wed Apr 10 18:04:07 CST 2013    v0.1
----------------------------------------------------------------
1. first beta vesion.
2. main function.


v2.0/main.cc
/*
 *  yerui@genomics.cn
 *  2012-12-31
 *  2013-04-09
 *
 */

#include "main.h"

int main( int argc, char **argv )
{
    float n_cut(0.1);
    float q_cut(0.5);
    char  qc_cut('F');
    ulong part_size(0);
    int   min_len(30);
    string bar_str(""); // UID type
    UIDt uid_t;
    int trim_h(0);

    mIntLong LenMap; // read length -> freqency


    int c;
    while ( (c=getopt(argc,argv,"n:q:c:p:m:u:t:h")) != -1 ) {
        switch (c) {
            case 'n': n_cut  = atof(optarg);    break;
            case 'q': q_cut  = atof(optarg);    break;
            case 'c': qc_cut = optarg[0];       break;
            case 'p': part_size = atoi(optarg); break;
            case 'm': min_len = atoi(optarg);   break;
            case 'u': bar_str = optarg;         break;
            case 't': trim_h  = atoi(optarg);   break;
            case 'h':
            default:  usage(argv[0]);           exit(1);
        }
    }

    cout << "# n: " << n_cut
         << "\tq: " << q_cut
         << "\tc: " << qc_cut
         << "\tp: " << part_size
         << "\tm: " << min_len
         << "\tu: " << bar_str
         << "\tt: " << trim_h
         << endl;

    int bar_size = bar_str.size();

    if ( bar_size ) {
        uid_t.assign( bar_str );
        min_len += bar_size * 2 + trim_h; // conside with adapter contamination
    }

    if ( argc - optind != 4 )
        usage(argv[0]), exit(1);

    char *f[4];
    for ( int i(0); i != 4; ++i ) {
        f[i] = argv[optind + i];
    }

    if ( access(f[0], R_OK) != 0 or access(f[1], R_OK) != 0 )
        cerr << "infile access error!" << endl, exit(1);

    igzstream in1( f[0] );
    igzstream in2( f[1] );

    string suffix(".fq.gz");
    string ofq1( f[2] ), ofq2( f[3] );

    if (   ofq1.rfind( suffix ) != ofq1.size() - 6
        or ofq2.rfind( suffix ) != ofq2.size() - 6 )
        cerr << "out fq files should have suffix as .fq.gz " << endl, exit(0);

    int part(1);
    string outpre1 = ofq1.substr(0, ofq1.size()-6 );
    string outpre2 = ofq2.substr(0, ofq2.size()-6 );

    if ( part_size > 0 ) {
        ofq1 = outpre1 + '.' + mtoa(part) + suffix;
        ofq2 = outpre2 + '.' + mtoa(part) + suffix;
    }

    ogzstream out1( ofq1.c_str() );
    ogzstream out2( ofq2.c_str() );

    if ( access(ofq1.c_str(), W_OK) != 0 or access(ofq2.c_str(), W_OK) != 0 )
        cerr << "output error!" << endl, exit(1);

    uint clean_rp(0);
    uint total_rp(0);

    uint adapter_c(0); // num of reads pairs trimed because of adapter
    uint short_c(0);   // num of reads pairs discarded because of adapter
    uint barcode_c(0); // num of reads pairs trimed because of barcode contamination
    uint lowqua_c(0);  // num of reads pairs discarded because of low quality
    uint fixseq_c(0);  // num of reads pairs discarded because of mismatch on fix_seq


    // fq1_raw  fq2_raw    fq1_clean  fq2_clean
    // 0        1          2          3

    ulong base_n[4] = {0,0,0,0};
    ulong N_n[4] = {0,0,0,0};
    ulong lq_n[4] = {0,0,0,0};  // low quality

    for (;;) {
        if ( !in1.good() or !in2.good() ) break;

        fq_read  a, b;
        in1 >> a;
        in2 >> b;

        ++total_rp;

        if ( a.id.size() < 1 ) {
            continue;
        }

        if ( !is_pair(a, b) ) {
            cerr << "ERROR: not pair reads in line: " << total_rp * 4 - 3 << endl;
            cerr << '[' << a.id << "]\t[" << b.id << ']' << endl;
            break;
        }

        uint a_len, b_len, a_N, b_N, a_low, b_low;

        a_len = a.seq.size();
        b_len = b.seq.size();
        a_N = a.base_n('N');
        b_N = b.base_n('N');
        a_low = a.low_qua_n(qc_cut);
        b_low = b.low_qua_n(qc_cut);

        base_n[0] += a_len;
        base_n[1] += b_len;
        N_n[0]    += a_N;
        N_n[1]    += b_N;
        lq_n[0]   += a_low;
        lq_n[1]   += b_low;

        // adapter contamination
        int ov_len = has_adapter(a,b);

        if ( ov_len >= min_len ) {
            a.trim(ov_len);
            b.trim(ov_len);

            ++adapter_c;
        }
        else if ( ov_len > 0 ) { // ov_len < min_len
            continue;
            ++short_c;
        }
        else { // ov_len == 0, no adapter contamination
            if ( bar_size ) {
                if ( barcode_clean(a, b, bar_size) ) // clean barcode contamination at the end of reads
                    ++barcode_c;
            }
        }

        if ( bar_size ) {
            if ( ! fetch_UID(a, b, uid_t, ov_len) ) {
                ++fixseq_c;
                continue;
            }
        }

        if ( trim_h > 0 ) {
            a.trim_head( trim_h );
            b.trim_head( trim_h );
        }

        a_len = a.seq.size();
        b_len = b.seq.size();
        a_N = a.base_n('N');
        b_N = b.base_n('N');
        a_low = a.low_qua_n(qc_cut);
        b_low = b.low_qua_n(qc_cut);


        if ( (float)a_N > a_len * n_cut or (float)b_N > b_len * n_cut ) {
            ++lowqua_c;
            continue;
        }

        if ( (float)a_low > a_len * q_cut or (float)b_low > b_len * q_cut ) {
            ++lowqua_c;
            continue;
        }

        ++LenMap[a_len];
        ++clean_rp;

        out1 << a;
        out2 << b;

        if ( part_size && clean_rp % part_size == 0 ) {
            ++part;
            ofq1 = outpre1 + '.' + mtoa(part) + suffix;
            ofq2 = outpre2 + '.' + mtoa(part) + suffix;

            out1.close();
            out1.clear();
            out1.open(ofq1.c_str());

            out2.close();
            out2.clear();
            out2.open(ofq2.c_str());
        }

        base_n[2] += a_len;
        base_n[3] += b_len;
        N_n[2]    += a_N;
        N_n[3]    += b_N;
        lq_n[2]   += a_low;
        lq_n[3]   += b_low;
    }

    in1.close();
    in2.close();
    out1.close();
    out2.close();

    cout << "\ntype\traw_data\tclean_data\n"
         << "Read_pairs\t" << total_rp << '\t' << clean_rp << '\n'
         << "Bases\t" << base_n[0] + base_n[1] << '\t' << base_n[2] + base_n[3] << '\n'
         << "N_in_fq1\t"  << N_n[0] << '\t' << N_n[2] << '\n'
         << "N_in_fq2\t"  << N_n[1] << '\t' << N_n[3] << '\n'
         << "Low_quality_base_in_fq1\t" << lq_n[0] << '\t' << lq_n[2] << '\n'
         << "Low_quality_base_in_fq2\t" << lq_n[1] << '\t' << lq_n[3] << "\n\n";

    printf("Rate(clean_base/raw)\t%.2f%%\n", (base_n[2]+base_n[3]) * 100.0 / (base_n[0]+base_n[1]) );
    printf("Rate(clean_read_pair/raw)\t%.2f%%\n\n", clean_rp * 100.0 / total_rp );

    cout << "Reason\tOperate\tReads_pairs\tRate_over_raw(%)\n";
    printf("adapter\ttrim\t%d\t%.2f\n", adapter_c, adapter_c * 100.0 / total_rp );
    printf("adapter\tdiscard\t%d\t%.2f\n", short_c, short_c * 100.0 / total_rp );
    printf("barcode\ttrim\t%d\t%.2f\n", barcode_c, barcode_c * 100.0 / total_rp );
    printf("low-quality\tdiscard\t%d\t%.2f\n", lowqua_c, lowqua_c * 100.0 / total_rp );
    printf("fix-seq-error\tdiscard\t%d\t%.2f\n\n", fixseq_c, fixseq_c * 100.0 / total_rp );

    cout << "# read_length\tfrequency\n";
    for ( mIntLong::iterator it = LenMap.begin(); it != LenMap.end(); ++it ) {
        cout << it->first << '\t' << it->second << endl;
    }

    exit(0);
}


v2.0/fq_read.h

// yerui@genomics.cn

// 2012-10-31  v1.0
// 2013-04-10  v1.1
// 2014-05-05  v2.0

#ifndef FQ_READ_H_
#define FQ_READ_H_

#include <iostream>
#include <string>
#include <set>

#include "UID_type.h"

using namespace std;

class fq_read {
public:
    string id;
    string seq;
    string qua;

    fq_read():id(""),seq(""),qua(""){}

    inline int base_n( char ) const;
    inline int low_qua_n( char ) const;
    inline void trim( int );
    inline void trim_head( int );

};

istream & operator >> ( istream &in,  fq_read &a )
{
    string plus;
    in >> a.id >> a.seq >> plus >> a.qua;
    return in;
}

ostream & operator << ( ostream &out, fq_read &a )
{
    out << a.id << '\n' << a.seq << "\n+\n" << a.qua << '\n';
    return out;
}

int fq_read::base_n( char c ) const
{
    int num(0);

    for ( int i(0); i != seq.size(); ++i )
        if ( seq[i] == c )  ++num;

    return num;
}

int fq_read::low_qua_n( char c ) const
{
    int num(0);

    for ( int i(0); i != qua.size(); ++i )
        if ( qua[i] < c )  ++num;

    return num;
}

void fq_read::trim( int ov_len )
{
    if ( seq.size() <= ov_len ) return;

    seq.erase( seq.begin() + ov_len, seq.end() );
    qua.erase( qua.begin() + ov_len, qua.end() );

    return;
}

void fq_read::trim_head( int len )
{
    seq.erase(0, len);
    qua.erase(0, len);
}


inline bool is_pair( const fq_read &a, const fq_read &b )
{
    int len = a.id.size();

    if ( len != b.id.size() )
        return false;
    else
        return a.id.substr(0, len-2) == b.id.substr(0, len-2);

}

string re_complement( const string &str )
{
    string a("");
    for ( int i = str.size()-1; i >= 0; --i ) {
        switch( str[i] ) {
            case 'A': a += "T"; break;
            case 'C': a += "G"; break;
            case 'G': a += "C"; break;
            case 'T': a += "A"; break;
            case 'N': a += "N"; break;
            default: cerr << "base error: [" << str << "]" << endl; break;
        }
    }

    return a;
}

void overlap(const string &seq1, const string &seq2, int p1, int p2, string &as, string &bs)
{
    if ( p1 < p2 ) {
        as = seq1.substr(0, seq1.length() - p2 + p1);
        bs = seq2.substr( p2-p1 );
    }
    else if ( p1 == p2 ) {
        as = seq1;
        bs = seq2;
    }
    else {
        as = seq1.substr( p1-p2 );
        bs = seq2.substr(0, seq1.length() - p1 + p2 );
    }
}

//  mismatch_num > n ? true : false

bool mismatch_num( string &x, string &y, int n )
{
    int len = x.size();
/*
    if ( len != y.size() )
        return false;
*/

    int mis_n(0);
    for ( int i(0); i != len; ++i ) {
        if ( x[i] != y[i] ) ++mis_n;
        if ( mis_n > n ) return true;
    }

/*
    if ( n == 2 ) {
        cout << "tail: " << x << "\ntail: " << y << endl;
    }
*/
    return false;
}


// return length of overlapped region

int has_adapter( fq_read &a, fq_read &b )
{

    int a_len = a.seq.size();

    int len(0); // length of overlapped region
    string rb = re_complement( b.seq );

    int s_len(10); // seed length
    set<int> df;   // record offsets between two reads seq
    set<int> okdf; // offsets that OK

    string as(""), bs("");

    // skip first base because its low quality
    for ( int i(1); i < 10; i += 2 ) {
        string seed = a.seq.substr(i, s_len);

        int start(i);
        for (;;) {
            size_t p = rb.find(seed, start);
            if ( p == string::npos) break;

            int offset = p - i;
            if ( df.find( offset ) != df.end() ) {
                start = p + 1;
                continue;
            }
            else
                df.insert( offset );

            overlap(a.seq, rb, i, p, as, bs);

            len = as.size();

            if ( mismatch_num(as, bs, 3) ) {
                start = p + 1;
                continue;
            }

            string tail_a = a.seq.substr(len, 10);
            string tail_b = b.seq.substr(len, 10);

            if ( mismatch_num(tail_a, tail_b, 2) ) {
                start = p + 1;
                continue;
            }

            okdf.insert( offset );
            if ( len > a_len / 2 ) return len;
            start = p + 1;
        }

        if ( len > a_len / 2 ) return len;
    }

    return ( okdf.size() > 0 ? a_len - *(okdf.begin()) : 0);
}

bool barcode_clean( fq_read &a, fq_read &b, int bar_size )
{
    int a_len = a.seq.size();
    string rb = re_complement( b.seq );

    int s_len(10); // seed length
    set<int> df;   // record offsets between two reads seq
    set<int> okdf; // offset that OK

    // skip first base because its low quality
    for ( int i(1); i < 10; i += 2 ) {
        string seed = rb.substr(i, s_len);

        int start(i);
        for (;;) {
            size_t p = a.seq.find(seed, start);
            if ( p == string::npos) break;

            int offset = p - i;
            if ( offset >= bar_size ) break;  // no contamination

            if ( df.find( offset ) != df.end() ) {
                start = p + 1;
                continue;
            }
            else
                df.insert( offset );

            string as(""), bs("");
            overlap(a.seq, rb, p, i, as, bs);

            int mismatch(0);
            if ( mismatch_num(as, bs, 3) ) {
                start = p + 1;
                continue;
            }

            okdf.insert( offset );
            break;
        }

        if ( okdf.size() )
            break;
    }

    if ( okdf.size() ) {
        int rd_len = a_len - bar_size + *(okdf.begin());

        a.trim(rd_len);
        b.trim(rd_len);
/*
        cout << "trim: " << a.seq << endl;
        cout << "trim: " << "           " << re_complement( b.seq ) << endl;
*/
        return true;
    }
    else
        return false;
}

bool fetch_UID( fq_read &a, fq_read &b, UIDt &v, int ov_len )
{
    int trim_len = v.bar_len + v.fix_len;

    // if fix_seq is not consistent, return false
    string a_fix = a.seq.substr(v.fix_pos, v.fix_len);
    string b_fix = b.seq.substr(v.fix_pos, v.fix_len);

    if ( mismatch_num(v.fix, a_fix, 1) or mismatch_num(v.fix, b_fix, 1) )
        return false;

    string uid_pr("#");

    uid_pr += a.seq.substr(v.bar_pos, v.bar_len);
    uid_pr += "#";
    uid_pr += b.seq.substr(v.bar_pos, v.bar_len);

    int insP = a.id.size() - 2;
    a.id.insert(insP, uid_pr);
    b.id.insert(insP, uid_pr);

    //cout << "test id\n" << a.id << '\n' << b.id << endl;

    if ( ov_len ) { // has adapter
        int rd_len = ov_len - trim_len * 2;

        a.seq = a.seq.substr(trim_len, rd_len);
        a.qua = a.qua.substr(trim_len, rd_len);

        b.seq = b.seq.substr(trim_len, rd_len);
        b.qua = b.qua.substr(trim_len, rd_len);
    }
    else {
        a.seq.erase(0, trim_len);
        a.qua.erase(0, trim_len);

        b.seq.erase(0, trim_len);
        b.qua.erase(0, trim_len);
    }

    return true;
}


#endif // FQ_READ_H_


v2.0/UID_type.h
// yerui@genomics.cn
// 2013-04-09

#ifndef UID_TYPE_H_
#define UID_TYPE_H_

#include <string>

using namespace std;

class UIDt
{
public:
    int bar_pos;
    int bar_len;
    int fix_pos;
    int fix_len;
    string bar;
    string fix;

    UIDt():bar_pos(0),bar_len(0),fix_pos(0),fix_len(0),bar(""),fix("") {};

    void assign( string &s );

};

void UIDt::assign( string &s )
{
    if ( s[0] == 'N' ) {  // barcode first

        size_t p = s.rfind("N");

        bar_pos = 0;
        bar_len = p + 1;

        fix_pos = bar_len;
        fix_len = s.size() - bar_len;
    }
    else {
        size_t p = s.find("N");

        bar_pos = p;
        bar_len = s.size() - p;

        fix_pos = 0;
        fix_len = p;
    }

    bar = s.substr(bar_pos, bar_len);
    fix = s.substr(fix_pos, fix_len);
}

ostream & operator << ( ostream &out, UIDt &a )
{
    out << a.bar << '\t' << a.bar_pos << '\t' << a.bar_len << '\n'
        << a.fix << '\t' << a.fix_pos << '\t' << a.fix_len << '\n';

    return out;
}

#endif // UID_TYPE_H_


v2.0/main.h
/*
 *  yerui@genomics.cn
 *  2012-10-31
 *  2013-04-09
 *  2014-05-08
 *
 */

#ifndef MAIN_H_
#define MAIN_H_

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>

#ifdef __GNUC__
#include <ext/hash_map>
#else
#include <hash_map>
#endif

#include "fq_read.h"
#include "gzstream.h"

using namespace __gnu_cxx;
using namespace std;

#define VERSION "v2.0"

typedef unsigned long ulong;
typedef unsigned int uint;
typedef map<int, ulong> mIntLong;

void usage(char *prog) {
    cout << "Program: fqclean   filtrate/trim pair-end reads with adapter or with too many low-quality-base\n"
            "Version: " << VERSION << "\n"
            "Contact: yerui <yerui@genomics.cn>\n\n"
            "Usage: " << prog << " in1.fq[.gz] in2.fq[.gz] out1.fq.gz out2.fq.gz\n\n"

            "       -n [f]    read in which rate of N > [f] will be discarded. [0.1]\n"
            "       -q [f]    read in which rate of low-quality-base > [f] will be discarded. [0.5]\n"
            "       -c [c]    phred quality < [c] is treated as low quality base. [ 'F' ]\n"
            "       -p [i]    [i] reads as a part to output. [0]\n"
            "       -m [i]    minimum length of reads after cleaning. [30]\n"
            "       -u [s]    UID types. ( NNNNNACT  means 5bp barcode + 3bp fix\n"
            "                              ACGNNNNN  meads 3bp fix + 5bp barcode\n"
            "                              NNNNNNNN  meads 8bp barcode. )\n"
            "       -t [i]    trim first N bp of reads (if has UID, trim N bp after UID) [0]\n"

            "\n" << endl;
}

template<typename M> string mtoa(M &i)
{
    ostringstream a;
    a << i;
    return a.str();
}

#endif // MAIN_H_
